woocommerce/woocommerce-android
# [Feature] Update Product Categories

Fixes #2186 

## Description
This PR help enables updating product categories. The categories were displayed on the product detail page in #2344. This PR enables the user to update or change the product categories from a list of available categories on the store.

This change goes in conjunction with the changes on FluxC here: wordpress-mobile/Wordpress-FluxC-Android#1572.

This step of the category changes to the app is special because it allows for a user to pull all categories from the server, which is a separate api, and merges the changes to the product model. You can follow more on the linked PR for changes in the library.

## UI Changes

Light Mode
<p float="left">
<img src="https://user-images.githubusercontent.com/1773060/81061474-c3c97100-8f17-11ea-80c6-1ff32c7b9e7e.png" width="300" />
<img src="https://user-images.githubusercontent.com/1773060/81061481-c6c46180-8f17-11ea-9cc4-8c441e219ff5.png" width="300" />
<img src="https://user-images.githubusercontent.com/1773060/81061491-cc21ac00-8f17-11ea-80b9-a09d08e20625.png" width="300" />
<img src="https://user-images.githubusercontent.com/1773060/81061496-cdeb6f80-8f17-11ea-8957-a0b7027b9b28.png" width="300" />
<img src="https://user-images.githubusercontent.com/1773060/81061507-d04dc980-8f17-11ea-91ff-33bcbc35623d.png" width="300" />
<img src="https://user-images.githubusercontent.com/1773060/81061509-d2178d00-8f17-11ea-96c1-963b8a4ee7a6.png" width="300" />
<img src="https://user-images.githubusercontent.com/1773060/81061515-d3e15080-8f17-11ea-8883-a67beb3dc49a.png" width="300" />
</p>

Dark Mode
<p float="left">
<img src="https://user-images.githubusercontent.com/1773060/81061530-db085e80-8f17-11ea-9d91-a5704e9e35c5.png" width="300" />
<img src="https://user-images.githubusercontent.com/1773060/81061535-dcd22200-8f17-11ea-9cbb-9a346f624fb8.png" width="300" />
<img src="https://user-images.githubusercontent.com/1773060/81061538-df347c00-8f17-11ea-8371-30ee031eeeff.png" width="300" />
<img src="https://user-images.githubusercontent.com/1773060/81061545-e0fe3f80-8f17-11ea-89ac-fb6cbdf2eb67.png" width="300" />
<img src="https://user-images.githubusercontent.com/1773060/81061551-e2c80300-8f17-11ea-9c59-72c734237604.png" width="300" />
<img src="https://user-images.githubusercontent.com/1773060/81061554-e491c680-8f17-11ea-9d76-f559294e2999.png" width="300" />
<img src="https://user-images.githubusercontent.com/1773060/81061558-e5c2f380-8f17-11ea-912a-1fe2f4a777fd.png" width="300" />
<img src="https://user-images.githubusercontent.com/1773060/81061562-e78cb700-8f17-11ea-96f0-3c5c7be3c518.png" width="300" />
</p>

## Event Changes
`PRODUCT_DETAIL_VIEW_CATEGORIES_TAPPED`
Fired when the categories section is clicked on the Product Detail tab.

`PRODUCT_CATEGORIES_SETTINGS_DONE_BUTTON_TAPPED`
Fired when the category selection is updated and the done button clicked.

`PRODUCT_CATEGORIES_LOADED`
Fired when the categories from the server are successfully loaded.

`PRODUCT_CATEGORIES_LOAD_FAILED`
Fired when the categories from the server failed to load.

`PRODUCT_CATEGORY_LIST_ITEM_TAPPED`
Fired when a product category is clicked on the product category list.

`PRODUCT_CATEGORIES_PULLED_TO_REFRESH`
Fired when categories are refreshed using pull or swipe gesture.

## Testing
- Sign into the app with the selected Woo store on your account
- Go to the Products Tab
- Select any Product, and you should be able to see the categories showing under the product detail.
- Clicking on the categories will take you to the `Product Categories Update` fragment, where you can select the categories where the product should be available.
- Click Done, and when you return to the Product Detail tab, you'll see the product draft has the new selection. You can click `Update` to update the categories on the Product.

## Author Checklist:

Update release notes:
